File: imagepipeline/src/test/java/com/facebook/imagepipeline/producers/SettableProducerContextTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.producers;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import com.facebook.imagepipeline.common.Priority;
import com.facebook.imagepipeline.request.ImageRequest;
import org.junit.*;
import org.junit.runner.*;
import org.mockito.*;
import org.robolectric.*;
import org.robolectric.annotation.*;

/**
 * Checks basic properties of NullProducer, that is that it always returns null.
 */
@RunWith(RobolectricTestRunner.class)
@Config(manifest= Config.NONE)
public class SettableProducerContextTest {
  @Mock public ImageRequest mImageRequest;
  private final String mRequestId = "mRequestId";
  private ProducerContextCallbacks mCallbacks1;
  private ProducerContextCallbacks mCallbacks2;
  private SettableProducerContext mSettableProducerContext;

  @Before
  public void setUp() {
    mSettableProducerContext = new SettableProducerContext(
        mImageRequest,
        mRequestId,
        mock(ProducerListener.class),
        mock(Object.class),
        ImageRequest.RequestLevel.FULL_FETCH,
        false,
        true,
        Priority.MEDIUM);
    mCallbacks1 = mock(ProducerContextCallbacks.class);
    mCallbacks2 = mock(ProducerContextCallbacks.class);
  }

  @Test
  public void testGetters() {
    assertEquals(mImageRequest, mSettableProducerContext.getImageRequest());
    assertEquals(mRequestId, mSettableProducerContext.getId());
  }

  @Test
  public void testIsPrefetch() {
    assertFalse(mSettableProducerContext.isPrefetch());
  }

  @Test
  public void testCancellation() {
    mSettableProducerContext.addCallbacks(mCallbacks1);
    verify(mCallbacks1, never()).onCancellationRequested();
    mSettableProducerContext.cancel();
    verify(mCallbacks1).onCancellationRequested();
    verify(mCallbacks1, never()).onIsPrefetchChanged();

    mSettableProducerContext.addCallbacks(mCallbacks2);
    verify(mCallbacks2).onCancellationRequested();
    verify(mCallbacks2, never()).onIsPrefetchChanged();
  }

  @Test
  public void testSetPrefetch() {
    mSettableProducerContext.addCallbacks(mCallbacks1);
    assertFalse(mSettableProducerContext.isPrefetch());
    mSettableProducerContext.setIsPrefetch(true);
    assertTrue(mSettableProducerContext.isPrefetch());
    verify(mCallbacks1).onIsPrefetchChanged();
    mSettableProducerContext.setIsPrefetch(true);
    // only one callback is expected
    verify(mCallbacks1).onIsPrefetchChanged();
  }

  @Test
  public void testSetIsIntermediateResultExpected() {
    mSettableProducerContext.addCallbacks(mCallbacks1);
    assertTrue(mSettableProducerContext.isIntermediateResultExpected());
    mSettableProducerContext.setIsIntermediateResultExpected(false);
    assertFalse(mSettableProducerContext.isIntermediateResultExpected());
    verify(mCallbacks1).onIsIntermediateResultExpectedChanged();
    mSettableProducerContext.setIsIntermediateResultExpected(false);
    // only one callback is expected
    verify(mCallbacks1).onIsIntermediateResultExpectedChanged();
  }

  @Test
  public void testNoCallbackCalledWhenIsPrefetchDoesNotChange() {
    assertFalse(mSettableProducerContext.isPrefetch());
    mSettableProducerContext.addCallbacks(mCallbacks1);
    mSettableProducerContext.setIsPrefetch(false);
    verify(mCallbacks1, never()).onIsPrefetchChanged();
  }

  @Test
  public void testCallbackCalledWhenIsPrefetchChanges() {
    assertFalse(mSettableProducerContext.isPrefetch());
    mSettableProducerContext.addCallbacks(mCallbacks1);
    mSettableProducerContext.addCallbacks(mCallbacks2);
    mSettableProducerContext.setIsPrefetch(true);
    assertTrue(mSettableProducerContext.isPrefetch());
    verify(mCallbacks1).onIsPrefetchChanged();
    verify(mCallbacks1, never()).onCancellationRequested();
    verify(mCallbacks2).onIsPrefetchChanged();
    verify(mCallbacks2, never()).onCancellationRequested();
  }
}

File: imagepipeline/src/test/java/com/facebook/imagepipeline/producers/SettableProducerContextTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.producers

import com.facebook.imagepipeline.common.Priority
import com.facebook.imagepipeline.core.ImagePipelineConfig
import com.facebook.imagepipeline.request.ImageRequest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/** Checks basic properties of SettableProducerContext. */
@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
class SettableProducerContextTest {
  @Mock private lateinit var imageRequest: ImageRequest

  @Mock private lateinit var config: ImagePipelineConfig

  private val requestId = "requestId"
  private lateinit var callbacks1: ProducerContextCallbacks
  private lateinit var callbacks2: ProducerContextCallbacks
  private lateinit var settableProducerContext: SettableProducerContext

  @Before
  fun setUp() {
    MockitoAnnotations.initMocks(this)
    settableProducerContext =
        SettableProducerContext(
            imageRequest,
            requestId,
            Mockito.mock(ProducerListener2::class.java),
            Mockito.mock(Any::class.java),
            ImageRequest.RequestLevel.FULL_FETCH,
            false,
            true,
            Priority.MEDIUM,
            config)
    callbacks1 = Mockito.mock(ProducerContextCallbacks::class.java)
    callbacks2 = Mockito.mock(ProducerContextCallbacks::class.java)
  }

  @Test
  fun testGetters() {
    assertThat(settableProducerContext.imageRequest).isEqualTo(imageRequest)
    assertThat(settableProducerContext.id).isEqualTo(requestId)
  }

  @Test
  fun testIsPrefetch() {
    assertThat(settableProducerContext.isPrefetch).isFalse()
  }

  @Test
  fun testCancellation() {
    settableProducerContext.addCallbacks(callbacks1)
    Mockito.verify(callbacks1, Mockito.never()).onCancellationRequested()
    settableProducerContext.cancel()
    Mockito.verify(callbacks1).onCancellationRequested()
    Mockito.verify(callbacks1, Mockito.never()).onIsPrefetchChanged()

    settableProducerContext.addCallbacks(callbacks2)
    Mockito.verify(callbacks2).onCancellationRequested()
    Mockito.verify(callbacks2, Mockito.never()).onIsPrefetchChanged()
  }

  @Test
  fun testSetPrefetch() {
    settableProducerContext.addCallbacks(callbacks1)
    assertThat(settableProducerContext.isPrefetch).isFalse()
    settableProducerContext.setIsPrefetch(true)
    assertThat(settableProducerContext.isPrefetch).isTrue()
    Mockito.verify(callbacks1).onIsPrefetchChanged()
    settableProducerContext.setIsPrefetch(true)
    // only one callback is expected
    Mockito.verify(callbacks1).onIsPrefetchChanged()
  }

  @Test
  fun testSetIsIntermediateResultExpected() {
    settableProducerContext.addCallbacks(callbacks1)
    assertThat(settableProducerContext.isIntermediateResultExpected).isTrue()
    settableProducerContext.setIsIntermediateResultExpected(false)
    assertThat(settableProducerContext.isIntermediateResultExpected).isFalse()
    Mockito.verify(callbacks1).onIsIntermediateResultExpectedChanged()
    settableProducerContext.setIsIntermediateResultExpected(false)
    // only one callback is expected
    Mockito.verify(callbacks1).onIsIntermediateResultExpectedChanged()
  }

  @Test
  fun testNoCallbackCalledWhenIsPrefetchDoesNotChange() {
    assertThat(settableProducerContext.isPrefetch).isFalse()
    settableProducerContext.addCallbacks(callbacks1)
    settableProducerContext.setIsPrefetch(false)
    Mockito.verify(callbacks1, Mockito.never()).onIsPrefetchChanged()
  }

  @Test
  fun testCallbackCalledWhenIsPrefetchChanges() {
    assertThat(settableProducerContext.isPrefetch).isFalse()
    settableProducerContext.addCallbacks(callbacks1)
    settableProducerContext.addCallbacks(callbacks2)
    settableProducerContext.setIsPrefetch(true)
    assertThat(settableProducerContext.isPrefetch).isTrue()
    Mockito.verify(callbacks1).onIsPrefetchChanged()
    Mockito.verify(callbacks1, Mockito.never()).onCancellationRequested()
    Mockito.verify(callbacks2).onIsPrefetchChanged()
    Mockito.verify(callbacks2, Mockito.never()).onCancellationRequested()
  }
}
